### PR TITLE
Simplify Tasty Trees 

### PR DESCRIPTION
**Pull request by odersky**
Several simplifications and fixes for tasty trees.

Also: Avoid crash if prefix of selection is higher-kinded
The test case caused a crash in NamedType, because `Lambda` is higher-kinded.
We now report an error instead.

### CODE REVIEW

File: tests/pos/TastyADT.scala [hunk 1]
```diff
@@ -121,6 +49,25 @@ object tasty {
...
 
+// ------ Definitions ---------------------------------
+
+  class Symbol {
+    def owner: Symbol = ???
+    def definition: Definition = ???
+  }
+  object NoSymbol extends Symbol
+
+  trait Definition {
```
Comment (nicolasstucki) [inline tests/pos/TastyADT.scala]: Should extend `Statement`

File: tests/pos/TastyADT.scala [hunk 2]
```diff
@@ -76,39 +71,39 @@ object tasty {
...
-    case Super(thiz: Term, mixin: Id | Empty)
-    case Typed(expr: Term, tpt: Term)
+    case TypeApply(fn: Term, args: List[TypeTree])
+    case Super(thiz: Term, mixin: Option[Id])
+    case Typed(expr: Term, tpt: TypeTree)
     case Assign(lhs: Term, rhs: Term)
     case Block(stats: List[Statement], expr: Term)
     case Inlined(call: Term, bindings: List[Definition], expr: Term)
-    case Lambda(method: Term, tpt: Term | Empty)
+    case Lambda(method: Term, tpt: Option[TypeTree])
```
Comment (smarter) [inline tests/pos/TastyADT.scala]: Why is tpt optional here?

File: tests/pos/TastyADT.scala [hunk 3]
```diff
@@ -202,6 +197,8 @@ object tasty {
     object ErasedImplicitMethodType extends SpecializedMethodTypeCompanion
 
     case class TypeBounds(loBound: Type, hiBound: Type)
+    case class NoPrefix()
+    object NoPrefix extends NoPrefix
```
Comment (smarter) [inline tests/pos/TastyADT.scala]: Why both a class and an object for NoPrefix?

Comment (nicolasstucki) [reply tests/pos/TastyADT.scala]: To allow `NoPrefix` to be used in `|` types and be a singleton at the same time.

File: tests/pos/TastyADT.scala (hunk 2 above)
Comment (nicolasstucki) [reply tests/pos/TastyADT.scala]: It should probably not be. https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/core/tasty/TastyFormat.scala#L88

Comment (nicolasstucki) [reply tests/pos/TastyADT.scala]: Actually, the doc is wrong. https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala#L1022-L1025

The docs should be updated to 
```
LAMBDA         Length meth_Term target_Type?
```

Comment (nicolasstucki) [reply tests/pos/TastyADT.scala]: Created #4285 to fix this

**Pull request merged**